### DISCUSSION TITLE
Security scopes

### DISCUSSION THREAD

**Discussion by mattlaue** [fastapi/fastapi#8228] [Questions] [answered]
**How can the scope parameter of the Security object be accessed?**

Security accepts __scopes__ as a parameter but the callable doesn't seem to be able to access it.  The use case is essentially what is mentioned in the documentation - require certain scopes to be present to access an endpoint, or generate a 403 error.

**What if you don't need the security parameter in the callback?**

In the above use case, I'd like to require one of a set of scopes to be present but which one isn't really important.   Using Security requires that a parameter be added like:
```
arg = Security(<callable>)
```
In callable(), test for the scopes and throw an HTTPException as needed.  The problem is that arg isn't needed, so its ugly to pass it to the function.  A much cleaner implementation would be to use a decorator - similar to the Starlette requires - like:

```
@requires(["user:read", "admin"])
```
Is this possible with the FastAPI design?  If so, how does the decorate get passed a list of scopes (from the request)?

**Comment by kthwaite** [comment]
This is a good question, and it isn't clear to me either. The [advanced dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/advanced-dependencies/) page in the docs would seem to suggest a pattern where you create a callable class or higher-order-function initialised from a list of scopes with a dependency on Security, and use that to check scope, e.g.

```python
class ScopeChecker:
    def __init__(self, scopes):
        self._scopes = scopes

    def __call__(self, request: Request, token: str = Security(oauth2_scheme)):
        token = validate_jwt_token(token)  # proper validation goes here
        if not token_has_required_scopes(token, self._scopes):
            raise HTTPException(403, detail='Forbidden')
        return token
```

but that either leads to a rather ugly syntax, or you have to create a global dep object for each set of scopes (!): 

```python
# either
@app.get('/')
async def root(request: Request, user: UserToken = Depends(ScopeChecker(['user']))):
    # ...

# or 

check_is_user = ScopeChecker(['user'])

@app.get('/')
async def root(request: Request, user: UserToken = Depends(check_is_user)):
    # ...
```

Based on my (cursory) reading of `fastapi.dependencies.utils` I've been subclassing `Depends` for something roughly like the following:

```python
class ScopedTo(Depends):
    """Dependency on particular scope.
    """
    def __init__(self, *scopes) -> None:
        super().__init__(self.__call__)
        self._scopes = scopes

    def __call__(self, request: Request, token: str = Security(oauth2_scheme)) -> TokenData:
        """Check scopes and return the current user.
        """
        token = validate_jwt_token(token) # proper validation goes here
        if not token_has_required_scopes(token, self._scopes):
            raise HTTPException(403, detail='Forbidden')
        return token
```

which can then be used like:

```python
@app.get('/')
async def root(request: Request, user: UserToken = ScopedTo('user', 'root:read')):
    # ...
```

But I'm not convinced that this is the best solution. I would welcome any clarification on this issue!

**Comment by tiangolo** [comment]
Sorry for the delay guys.

You're right, currently, there's no way to access the scopes defined in a `Security` inside of its dependency/security function. And it totally makes sense to be able to access them.

I'm thinking about implementing it as an additional parameter that you can declare in the security function, I imagine a class `fastapi.security.Scopes` that has an attribute `scopes` with the list of scopes declared.

So, you would define your dependency something like:

```Python
async def get_current_user(scopes: Scopes, token: str = Security(oauth2_scheme)):
    for scope in scopes.scopes:
        ...
```

It has to be a predefined class instead of a plain list to be able to recognize it and solve it while solving the dependencies. But I think that would be the simplest/cleanest way to get the scopes inside your dependencies without breaking or compromising anything else.

Would that sound good for you?

**Comment by mattlaue** [comment]
This implementation is almost exactly the same as what we implemented for our application.  I'd suggest something like (if possible):

```python
async def get_current_user(scopes: Scopes(Security(oauth2_scheme))):
    for scope in scopes.scopes:
        ...
```

This way you don't have an unused parameter `token` cluttering the function.

**Comment by tiangolo** [comment, marked as answer]
Here it is! :tada: :cake: :rocket: 

https://fastapi.tiangolo.com/tutorial/security/oauth2-scopes/

---

@mattlaue the problem is that you probably still want to read the token, at least at some point, to verify that it's valid.

But the new `SecurityScopes` allows you to read the scopes in a dependency that can be used in other dependencies or path operations. 

Those dependencies and path operations can declare their own required scopes. And the sub-dependency (that is probably reading the token and verifying it directly) can access all those required scopes from the *dependants*.

---

So, you can have a central point that checks and verifies all the scopes, and then in different *path operations* you can have `Security` dependencies with different scopes, for example:

```Python
@app.get("/somepath1/")
def some_path_1(current_user: User = Security(get_current_user, scopes=["some", "scopes"])):
    ...

@app.get("/somepath2/")
def some_path_2(current_user: User = Security(get_current_user, scopes=["something", "different"])):
    ...

def other_dependency(current_user: User = Security(get_current_user, scopes=["subdependency", "required"])):
    ...

@app.get("/somepath3/")
def some_path_3(sub_dep = Depends(other_dependency), current_user: User = Security(get_current_user, scopes=["more"])):
    ...
```

**Comment by tiangolo** [comment]
I assume this is solved now, so I'll close this issue. But feel free to add more comments or create new issues.

**Comment by lsaint** [comment]
Is there any mechanism to overwrite security scopes?
I had 10 URL in a router:

```
router = Router(dependencies=[Security(XXXBearer(), scopes=["admin"])])

@router.post("/url1/")
def url1_func():
    ...

@router.post("/url2/")
def url2_func():
    ...

@router.post("/url3/")
def url2_func():
    ...

...
```


let's say, the url3's scopes only require `["user"]`,  
how should I do?

```
@router.post("/url3/")
def url2_func(a = Security(XXXBearer(), scopes=["user"])):
    # it did not work like this
    # the require scopes were still ["admin"]
```

**Reply by megamanhxh** [reply]
Does `use_cache=False` gives the same effect?
```python
def url2_func(a = Security(XXXBearer(), scopes=["user"], use_cache=False)):
```

**Comment by megamanhxh** [comment]
Is there a way to extend the SecurityScopes to support other parameters like `permissions` or any claim base attributes?

The current use-case is creating a custom security that allow me to control based on function checking mechanics.

```python
@router.get("/me", response_model=User)
async def get_all_user(
        current_user: Annotated[
            User,
            Security(get_current_user, scopes=["me", "permissions"], permissions=["manage_user"])
        ]
):
    # Return all users
    pass
```

Where `get_current_user()` is the following:

```python
async def get_current_user(
        security_scopes: SecurityScopes,
        token: Annotated[str, Depends(oauth2_scheme)],
        company_id: Annotated[int, Depends(company_id_auth)],
        db: Session = Depends(get_db),
):
    ...

    if security_scopes.scopes:
        ... # cross check the scopes against the token

    if security_scopes.permissions: # Exception! permissions is not part of SecurityScopes
        ... # cross check the permissions against the token

    ... # Own validation
    user = get_user(db, username=token_data.username)
    ... # Cross validation
    return user
```

Since `security_scopes` will always based on the class `SecurityScopes`, it will only going to take the parameter `scopes`. Is there any other way to pass and get custom attribute like dictionary of objects allowing me to do my way of assigning security scopes as permissions while maintaining the `scopes` too?

```python
async def get_all_user(
        current_user: Annotated[
            User,
            Security(get_current_user, scopes=["me", "permissions"], extra={"permissions": ["manage_user"]})
        ]
):
```

I need to change my approach to utilize scope as permission system in additional of authorizing resources.

**Reply by megamanhxh** [reply]
Upon looking at the source code, it will always will use `SecurityScopes` as a class to utilize `security_scopes` dependent.
https://github.com/tiangolo/fastapi/blob/a3f1689d78abef3d5d54ba655ec87cef81bf7384/fastapi/dependencies/utils.py#L630-L632

What `SecurityScopes` offer is just focusing on one parameter:
https://github.com/tiangolo/fastapi/blob/a3f1689d78abef3d5d54ba655ec87cef81bf7384/fastapi/security/oauth2.py#L228-L231

However, since `SecurityScopes` is being constructed as a part of dependencies in the code, there is no way to override it or change it during the runtime.

Which I wished to be able to do the following:
```python
from fastapi.security import SecurityScopes

class SecurityPermissionsScopes(SecurityScopes):

    def __init__(self, scopes: Optional[List[str]] = None, permissions: Optional[List[str]] = None):
        super().__init__(scopes=scopes)
        self.permissions = permissions or []
        self.permission_str = " ".join(self.permissions)
```
```python
class SecurityPermissions(Security):
    def __init__(
            self,
            dependency: Optional[Callable[..., Any]] = None,
            *,
            scopes: Optional[Sequence[str]] = None,
            permissions: Optional[Sequence[str]] = None,
            use_cache: bool = True,
    ):
        super().__init__(dependency=dependency, use_cache=use_cache)
        self.permissions = permissions or []
        self.scopes = scopes or []
```
```python
def SecurityPermissions(
        dependency: Optional[Callable[..., Any]] = None,
        *,
        scopes: Optional[Sequence[str]] = None,
        permissions: Optional[Sequence[str]] = None,
        use_cache: bool = True,
) -> Any:
    return security_param.SecurityPermissions(
        dependency=dependency,
        scopes=scopes,
        permissions=permissions,
        use_cache=use_cache
    )
```

Is there any workaround to allow dynamically assign the attributes? or any suggestion to mitigate the problem without relying on scopes entirely?

**Reply by alexandru0-dev** [reply]
I after a bit of fiddling around I got to the same results as you and was disappointed of how I couldn't use custom attributes
My use case was to force to custom StrEnums for type checking 
@tiangolo should I create a new issue for this?

**Comment by eblaas** [comment]
I worked around this problem by creating a class which extends "str", so you can pass through every additional data u want

```python
class ScopePredicate(str):
    def __init__(self, scopes: list[str]):
        self.scopes = scopes
       self.any_other_field = ..

    def __new__(cls, scopes: list[str]):
        value = f" {cls.__name__.lower()} ".join(tuple(sorted(set(scopes))))
        member = str.__new__(cls, value)
        member._value_ = value
        return member

class And(ScopePredicate):
    pass

class Or(ScopePredicate):
    pass


def get_current_user(
    security_scopes: SecurityScopes
) -> UserData:
    for scope in security_scopes.scopes:
        if isinstance(scope, Or):
            # access any other field  scope.any_other_field
           pass
        elif isinstance(scope, And):
           pass
    return UserData()
           
           
Annotated[
    UserData,
    Security(get_current_user, scopes=[Or(["entitlement-a", "entitlement-b"])]),
]

```
